torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+                self.optim.add_param_group(param_groups[-1])
+
+    ##########################
+    # Private helper functions
+    ##########################
+
+    def _broadcast_params(self) -> None:
+        """Helper function to broadcast all the parameters from a given device"""
+
+        with torch.no_grad():
+            # All the params on this device (inc all ranks)
+            for (
+                device,
+                device_params,
+            ), bucket_list in zip(self.per_device_params.items(), self._buckets.values()):
+
+                # Go through all the ranks and broadcast the new parameters, bucket the smallest ones
+                for (src_rank, params), bucket in zip(enumerate(device_params), bucket_list):
+
+                    global_src_rank = self.get_global_rank(self.group, src_rank)
+
+                    for param in params:
+                        # Bucket broadcast
+                        if self.should_bucket_param[param]:
+                            assert bucket.append(param), "Bucket overflow: max %s - current %s - adding %s" % (
+                                bucket.max_size,
+                                bucket.current_offset,
+                                param.numel(),
+                            )
+
+                            if bucket.full():
+                                self.work_handles.append(
+                                    _Workhandle(
+                                        handle=dist.broadcast(
+                                            tensor=bucket.buffer, src=global_src_rank, group=self.group, async_op=True
+                                        ),
+                                        callback=bucket.unroll,
+                                    )
+                                )
+
+                        # Direct
+                        else:
+                            self.work_handles.append(
+                                _Workhandle(
+                                    handle=dist.broadcast(
+                                        tensor=param.data, src=global_src_rank, group=self.group, async_op=True
+                                    ),
+                                    callback=None,
+                                )
+                            )
+
+    def _broadcast_state_dict(self) -> None:
+        """Broadcast this rank's state shard, discard others"""
+        empty_buffer = torch.tensor([0], dtype=torch.uint8, device=self._device)
+
+        for rank in range(self.world_size):
+            if rank == self.rank:
+                # Send the state to the reference replica
+                logging.debug(
+                    "Sending the sharded optimizer state to the reference replica from rank %s",
+                    rank,
+                )
+                _broadcast_object(
+                    self.local_state_dict(), src_rank=self.global_rank, group=self.group, dist_device=self._device
```

Comment:
agreed, but this part needs to be rewritten anyway (ideally using remote references), so we might as well directly do that

Reply:
Are we planning to rewrite it in this PR or a follow up PR?

Reply:
Personally I would have preferred everything RPC based, *if* it's possible to have a single c10/rpc init (ie. users do not have anything else to do, they init c10 as usual, use this, and when checkpointing they get the state properly), it would be a lot cleaner to my eyes. Now I also think that there could be value in having a first version and iterating, i think it was @mrshenli thinking, so no very strong feelings.